govcd/tm_ldap.go

```diff
@@ -15,40 +15,138 @@ import (
 	"strings"
 )
 
-// TmLdapConfigure configures LDAP for the Tenant Manager "System" organization
-func (vcdClient *VCDClient) TmLdapConfigure(settings *types.TmLdapSettings) (*types.TmLdapSettings, error) {
+// TmLdapConfigure configures LDAP for the Tenant Manager "System" organization. If trustSslCertificate=true,
+// it will automatically trust the certificate of LDAP server, only if settings.IsSsl=true.
+// If settings.IsSsl=true and trustSslCertificate=false this method returns an error
+func (vcdClient *VCDClient) TmLdapConfigure(settings *types.TmLdapSettings, trustSslCertificate bool) (*types.TmLdapSettings, error) {
+	var trustedCert *TrustedCertificate
+
 	if !vcdClient.Client.IsTm() {
 		return nil, fmt.Errorf("this method is only supported in TM")
 	}
+	if settings == nil {
+		return nil, fmt.Errorf("LDAP settings cannot be nil")
+	}
+
+	// Always test connection, because this method always changes LDAP. There's no NONE mode like in
+	// types.OrgLdapSettingsType
+	connResult, err := ldapValidateConnection(&vcdClient.Client, settings.HostName, settings.Port, settings.IsSsl)
+	if err != nil {
+		return nil, err
+	}
+
+	if !trustSslCertificate && connResult.TargetProbe.SSLResult == types.UntrustedCertificate {
+		return nil, fmt.Errorf("cannot configure LDAP '%s:%d' over SSL without trusting the SSL certificate", settings.HostName, settings.Port)
+	}
+
+	if trustSslCertificate && connResult.TargetProbe.SSLResult == types.UntrustedCertificate {
+		// We retrieve existing LDAP configuration to check if the given input settings are for a fresh connection, or to override
+		// an existing one
+		existing, err := vcdClient.TmGetLdapConfiguration()
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving existing LDAP configuration: %s", err)
+		}
+
+		// If SSL is configured and retrieved settings are empty, or the hostname changed, we need to trust the certificate
+		needsCheckCert := settings.IsSsl && (existing.HostName == "" || existing.HostName != settings.HostName)
```

Comment:
As far as I saw, if you change the host in UI, it doesn't re-trigger the certificate trusting popup. I think it should? So not sure if this should be so "smart" or simplify